### ISSUE TITLE
False negative: missing/tab spacing between type and variable not fully detected

### ISSUE THREAD

**Issue by Tieshiyang** [checkstyle/checkstyle#17728]
I have read check documentation: https://checkstyle.sourceforge.io/checks/whitespace/genericwhitespace.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

From: https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace
> Beyond where required by the language or other style rules, and apart from literals, comments and Javadoc, a single ASCII space also appears in the following places only.
• Between the type and variable of a declaration.

Code: 
```java
import java.util.List;

class Test {
  boolean	withTab1; // Expected 2 Violations for TAB and no space

  List<String>items; // Expected 1 Violation for no space
  List<String>	withTab2; // Expected 2 Violations for TAB and no space

  int[]numbers; // Expected 1 Violation for no space
  String[][]	withTab3; // Expected 2 Violations for TAB and no space
}

```

Cli: 
```cmd
$ java -jar checkstyle-10.26.1-all.jar -c google_checks.xml Test.java 
Starting audit...
[WARN] Test.java:5:10: Line contains a tab character. [FileTabCharacter]
[WARN] Test.java:7:14: GenericWhitespace '>' should followed by whitespace. [GenericWhitespace]
[WARN] Test.java:8:15: Line contains a tab character. [FileTabCharacter]
[WARN] Test.java:11:13: Line contains a tab character. [FileTabCharacter]
Audit done.
```

Formatter:
```cmd
$ java -jar google-java-format-1.28.0-all-deps.jar Test.java > FormattedCode.java
$ diff -Naru Test.java FormattedCode.java
--- Test.java   2025-09-02 10:55:55.998605900 +0800
+++ FormattedCode.java  2025-09-02 10:56:09.892189400 +0800
@@ -1,13 +1,11 @@
 import java.util.List;
-import java.util.Map;

 class Test {
-  boolean      withTab1; // Expected 2 Violations for TAB and no space
+  boolean withTab1; // Expected 2 Violations for TAB and no space

-  List<String>items; // Expected 1 Violation for multiple spaces
-  List<String> withTab2; // Expected 2 Violations for TAB and no space
-
-  int[]numbers; // Expected 1 Violation for no space
-  String[][]   withTab3; // Expected 2 Violations for TAB and no space
+  List<String> items; // Expected 1 Violation for multiple spaces
+  List<String> withTab2; // Expected 2 Violations for TAB and no space

+  int[] numbers; // Expected 1 Violation for no space
+  String[][] withTab3; // Expected 2 Violations for TAB and no space
 }
```

---

According to Google Style, there must be exactly one space between the type and the variable.

However 
* For cases like `int[]numbers`, where there is no space at all between the array type and the variable, there is currently no check that reports this violation.

* When a TAB is used instead of a space, Checkstyle only reports `FileTabCharacter`, and this seems to suppress the expected whitespace violation. This behavior appears in several checks, and might be by design — but it means the missing-space issue is not flagged.

As a result, both TAB instead of space and no space at all become false negatives in type–variable spacing.

---

**Comment by romani** [comment]
https://checkstyle.sourceforge.io/checks/whitespace/filetabcharacter.html#Notes

> When the FileTabCharacter check is used with the default configuration, only the first instance of a tab character is reported.

so  this is expected.


-----------

>int[]numbers; 

this is valid false-negative

**Comment by VeniceWish** [comment]
Thanks for the reply and approving this issue!
But I want to clarify that my point is not actually about reporting multiple TAB errors, in fact.

It was my mistake — the variable name `withTab2` just meant it's the second example with a TAB.
I realize now that this probably caused some confusion, sorry about that!

Example:
```java
List<String>items;      // Expected 1 violation for no space (line 7)
List<String>	withTab2; // Expected 2 violations for TAB and no space (line 8)
```

Current result:
```bash
[WARN] Test.java:7:14: GenericWhitespace '>' should followed by whitespace. [GenericWhitespace]
[WARN] Test.java:8:15: Line contains a tab character. [FileTabCharacter]
```

As you can see, line 8 only reports the TAB, and the missing space violation in `GenericWhitespace` is skipped.
I’m not sure if reporting both is necessary, but I raised it because it seems TAB is treated as whitespace, so the `GenericWhitespace` is ignored.

**Comment by romani** [comment]
please create separate issue on this, fix will be different, so better keep issues separately to not confuse future contributor who will try to fix it.

**Comment by stoyanK7** [comment]
I am on it.

**Comment by stoyanK7** [comment]
I'm wondering how to approach this.

Adding the `RBRACK` token to `WhitespaceAfterCheck` is possible, but there are many cases like `+=`, `++`, bitwise operators, etc.. to take into account.

**Comment by stoyanK7** [comment]
@romani

There are lots of cases to cover with brackets and it's not a simple matter of adding `RBRACK` to `WhitespaceAfterCheck`.
```java
arr[] nums          // whitespace
arr[i] += 1          // whitespace
arr[i]++             // no whitespace
arr[i].foo()         // no whitespace
arr[i][j] = 2       // no whitespace after first, but after second
int arr[];             // c-style array definitions?
List<String[]>   // no whitespace after
...                    // whatever else I cannot think of but will
                       // pop up with checkstyle-tester
```

How about creating a new check called `BracketWhitespace` similar to [`GenericWhitespace`](https://checkstyle.sourceforge.io/checks/whitespace/genericwhitespace.html#GenericWhitespace). Do you see this as a good approach?

**Comment by romani** [comment]
You are right, extension of Check is error prone.

```
int[]numbers;
List<String>items;
```

What if we focus on it from different angle. `Between the type and variable of a declaration.` And we will have Check focus on this target, without details on symbols that type is using.

There will be some overlap by Checks, but it is ok.

**Comment by stoyanK7** [comment]
In other words, add [`TYPE` token](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#TYPE) to WhitespaceAfterCheck?

**Comment by romani** [comment]
Interesting idea.
https://checkstyle.org/checks/whitespace/whitespaceafter.html#tokens

It worth to try .

If any false positives possible, let's abandon extension of existing and simply create new check for simple goal.

**Comment by AryanKadam0007** [comment]
I would like to contribute to this issue.

The Google style checker currently reports a false negative for missing spacing between a type and a variable. This spacing violation is not being detected in certain cases, which causes inconsistent formatting checks.

I am planning to analyze the root cause, add a reproducible test case, and submit a fix to ensure the spacing rule is correctly enforced.

Please let me know if this issue is still open for contribution or if there are any specific guidelines I should follow before submitting a PR.

**Comment by romani** [comment]
this will be covered by new Check - [ArrayBracketWhitespace #18064](https://github.com/checkstyle/checkstyle/issues/18064) 

in scope of this issue we will need to activate such new Check in google_checks.xml

**Comment by romani** [comment]
new Check is merged, now we are ready to add it google style

**Comment by romani** [comment]
@Atharv3221 , please help us ti make it

**Comment by romani** [comment]
Fix is merged
